velox/exec/HashBuild.h

```diff
@@ -18,35 +18,103 @@
-  // processing all the input and without finishing building the hash table.
+  /// Represents the result of HashBuild operators: a hash table, an optional
+  /// restored spill partition id associated with the table, and the spilled
+  /// partitions while building the table if not empty. In case of an anti join,
+  /// a build side entry with a null in a join key makes the join return
+  /// nothing. In this case, HashBuild operators finishes early without
+  /// processing all the input and without finishing building the hash table.
   struct HashBuildResult {
+    HashBuildResult(
+        std::shared_ptr<BaseHashTable> _table,
+        std::optional<SpillPartitionId> _restoredPartitionId,
+        SpillPartitionIdSet _spillPartitionIds)
+        : antiJoinHasNullKeys(false),
+          table(std::move(_table)),
+          restoredPartitionId(std::move(_restoredPartitionId)),
+          spillPartitionIds(std::move(_spillPartitionIds)) {}
+
+    HashBuildResult() : antiJoinHasNullKeys(true) {}
+
+    bool antiJoinHasNullKeys{false};
     std::shared_ptr<BaseHashTable> table;
-    bool antiJoinHasNullKeys;
+    std::optional<SpillPartitionId> restoredPartitionId;
+    SpillPartitionIdSet spillPartitionIds;
+  };
+
+  std::optional<HashBuildResult> tableOrFuture(
+      ContinueFuture* FOLLY_NONNULL future);
+
+  /// Invoked by HashProbe operator after finishes probing the built table to
+  /// set one of the previously spilled partition to restore. The HashBuild
+  /// operators will then build the next hash table from the selected spilled
+  /// one.
+  bool setSpillInput();
+
+  /// Contains the spill input for one HashBuild operator: a shard of previously
+  /// spilled partition data. 'spillPartition' is null if there is no more spill
+  /// data to restore.
+  struct SpillInput {
+    explicit SpillInput(
+        std::unique_ptr<SpillPartition> spillPartition = nullptr)
+        : spillPartition(std::move(spillPartition)) {}
+
+    std::unique_ptr<SpillPartition> spillPartition;
   };
 
-  std::optional<HashBuildResult> tableOrFuture(ContinueFuture* future);
+  /// Invoked by HashBuild operator to get one of previously spilled partition
+  /// shard to restore. The spilled partition to restore is set by HashProbe
+  /// operator after finishes probing on the previously built hash table.
+  /// If HashBuild operator calls this early, 'future' will be set to wait
+  /// asynchronously. If there is no more spill data to restore, then
+  /// 'spillPartition' will be set to null in the returned SpillInput.
+  std::optional<SpillInput> spillInputOrFuture(
+      ContinueFuture* FOLLY_NONNULL future);
 
  private:
-  std::shared_ptr<BaseHashTable> table_;
-  bool antiJoinHasNullKeys_{false};
+  uint32_t numBuilders_{0};
+
+  std::optional<HashBuildResult> buildResult_;
+
+  // If not null, set to the currently restoring spill partition id.
+  std::optional<SpillPartitionId> restoringSpillPartitionId_;
```

Comment:
Just to clarify, restoringSpillPartitionXxx member variables are populated by the bridge itself. When probe side finished processing, the bridge picks the first partition from spillPartitionSets_, splits it into "even" shards among the HashBuild operators and notifies these operators that they can start building HashTables from these shards.